### PR TITLE
WIP: clippy cleanup

### PR DESCRIPTION
**Pull request by imp**
Clippy has quite a lot to say about the code base.
This intended to be a gradual cleanup.
Feel free to reject if there are other ideas or policies on clippy, or pickup up commits incrementally if deemed useful.
I'll be adding more work here as time permits.

### CODE REVIEW

**Comment by djc** [comment]
@imp thanks for the PR! I still had a branch with some clippy-suggested fixes sitting in my local clone, so I've pushed that out to #104 to prevent duplicate work. If you want, you can cherry-pick my commit into your branch?

**Comment by imp** [comment]
@djc 👍 Sure thing!

**Comment by djc** [comment]
Also, I have run clippy on the code base a few times and cleaned stuff up. However, in doing so I've mostly ignored `clippy::cast_lossless` for now, as I'm not quite convinced of the cost/benefit there. I also feel `clippy::needless_pass_by_value` often has false positives in the sense that we sometimes want to pass by value conceptually even if the callee doesn't technically need ownership. `too_many_arguments` and `cyclomatic_complexity` I've ignored for now as we work out better APIs and merge/split functions to improve the internal design. `option_option` does need work, but definitely needs @Ralith's input for better-defined enums.

I think that covers everything I know about but have not fixed (or postponed) so far.

**Comment by imp** [comment]
Quite agree with your points. Let me see if I can come up with anything simple and non-controversial.

**Comment by djc** [comment]
Perhaps we should also add local `allow()` for the `needless_pass_by_value`, that would be a good addition IMO.

File: quinn-proto/src/stream.rs [hunk 1]
```diff
@@ -91,12 +91,11 @@ impl Send {
             }
             | SendState::ResetRecvd {
                 ref mut stop_reason,
-            } => match stop_reason.take() {
-                Some(error_code) => {
+            } => {
+                if let Some(error_code) = stop_reason.take() {
```
Comment (djc) [inline quinn-proto/src/stream.rs]: Does `cargo fmt` mandate the surrounding braces here? It seems like it could do without them.

Comment (imp) [reply quinn-proto/src/stream.rs]: Yep, I was surprised too, but rustfmt insisted extra curly braces are required. Go figure.

**Comment by imp** [comment]
>  However, in doing so I've mostly ignored `clippy::cast_lossless` for now, as I'm not quite convinced of the cost/benefit there.

I personally stopped using ` as ` conversion in my own code in favor of `uXX::from()`.
The latter is essentially the former:
```rust
impl From<u8> for u64 {
    #[inline]
    fn from(smalr: u8) -> u64 {
        small as u64
    }
}
```
However, it is a matter of policy. If ` as ` conversion if preferred though, I'd suggest setting
 `#![allow(clippy::cast_lossless)]` at the top level.
Thoughts?

**Comment by djc** [comment]
Can you fix up my formatting bug in the initial commit? That way CI will pass.

**Comment by imp** [comment]
> Can you fix up my formatting bug in the initial commit? That way CI will pass.

Rebased on latest `master`, which eliminated the offending code.

**Comment by imp** [comment]
Commits ed66828 and 6817993 are not clippy-related.
First a style cleanup, which is admittedly a matter of personal preferences. If this is acceptable I suggest conversion throughout the code base.
Second is a usability improvement, prompted by the follow-up commit.

**Comment by djc** [comment]
I'm not convinced about adding `is_client()` and `is_server()`. It is slightly more concise, but at the expense of adding boilerplatey functions to the enum's impl block.

Collapsing the `if` can be done even without the `is_client()` and `is_server()` change, I assume.

**Comment by imp** [comment]
@djc Correct, collapsing `if` does not require `is_client()` change.

I'd personally prefer concise over the slim impl block, but that's ultimately your call.

**Comment by djc** [comment]
@Ralith do you have an opinion either way on `is_client()`/`is_server()`?

**Comment by Ralith** [comment]
I'm okay with it; I'm generally on board with going to small amounts of extra effort to make downstream code read better.

**Pull request merged**